---
title: "DNM process June 2022"
output:
  html_document: default
---

## INITIALIZE
### libraries
```{r}
options(scipen=999)

library(here)
library(data.table)
library(stringr)
library(ggplot2)
library(ggrepel)
library(splitstackshape)
library(gridExtra)
library(ggbeeswarm)
library(gridGraphics)
library(grid)
library(gtable)
library(ggpubr)
library(cowplot)
library(scales)
#library(igraph)
library(RColorBrewer)
library(ggridges)
library(ggsci)
library(wesanderson)
library(viridis)
#library(VennDiagram)
library(tidyr)
library(dplyr)
library(MASS)
library(png)
require(plotrix)

```

### get posterior densities
```{r}

aut_post=read.table("ref_tables/autosome_hs.summary_stats.7_15.tsv",header=T)
aut_post$group = aut_post$group2 = "Autosome"

for (i in 1:6){
test=fread(paste0("posterior_dists/posterior_densities.",i,".txt.gz"), header=T)[,-1]
assign(paste0("test",i), as.list(test))
}
yax=c(test1, test2, test3, test4, test5, test6)
yax = yax[names(yax) %in% aut_post$Gene]
saveRDS(yax, "ref_tables/yax_aut.Rda")

```
### coverage-based and other exclusions -- src files
```{r}
aut_post=read.table("ref_tables/autosome_hs.summary_stats.7_15.tsv",header=T)
aut_post$group = aut_post$group2 = "Autosome"
par_post = fread("ref_tables/par.summary_stats.6_9.tsv")
par_post$group = par_post$group2 = "X (PAR)"
y_homolog_gene_list = fread("ref_tables/npx_posteriors.9_13.txt")[,1]
colnames(y_homolog_gene_list) = "Gene"
y_homolog_gene_list$group2 = "X (Y-homolog, non-PAR)"
npx_post = read.table("ref_tables/avg_x.stats.6_1_22.tsv", header=TRUE)[,c("Gene","avg_hs_log10_ci_low", "avg_hs_log10_ci_high","avg_hs_log10_map")]
colnames(npx_post) = c("Gene","log10_ci_low", "log10_ci_high","log10_map")
npx_post$group = "non-par X"
x_post = merge(npx_post, y_homolog_gene_list, by="Gene", all.x=T)
x_post$group2[is.na(x_post$group2)] = "X (Other non-PAR)"
x_post = rbind(x_post, par_post)
all_post = rbind(x_post, aut_post)
post = all_post
post$SYMBOL=post$Gene

prob = fread("ref_tables/problematic_genes.11_19.csv", header=T)
prob = rbind(prob, data.frame(gene=c("C2orf15", "C2ORF15","C1orf220","C1ORF220"),flag=c("duplicated","duplicated","duplicated","duplicated")))

gnomad = fread("ref_tables/gnomad.v2.1.1.lof_metrics.by_gene.txt", header=T)
gnomad$SYMBOL = gnomad$gene
gnomad$Feature = gnomad$transcript

coverage = fread("ref_tables/coverage_sensitive_genes.txt", header=T)
coverage$gene = coverage$SYMBOL

test=merge(coverage, gnomad, by=c("chromosome","gene","Feature"), all.y=T)
test=merge(prob, test, by="gene", all.y=T)
test$flag[is.na(test$flag)] = "none"
test = test[!gene %in% prob$gene[prob$flag=="duplicated"]]

test$flag2 = ifelse(!is.na(test$mu_lof),"ok","no_lof")
test = test[flag2=="ok"] #507 genes removed

test$frac = (test$count_HC_postfilter)/test$count_HC
test$flag3 = ifelse(test$AF_nfe_filtered>0 & test$frac<.5,1,0)
test = test[!test$flag=="AN_too_low"]
coverage_sens_genes = as.character(c(test$gene[test$flag3==1], prob$gene[prob$flag=="duplicated"]))
test = test[test$flag3==0]

excl.list = test[flag %in% c("wrong_model","wrong_model_x")][,"gene"]
write.table(excl.list, "out/Supplementary_file_1.txt",quote=F, sep="\t", row.names=F, col.names=T)
test = test[!flag %in% c("wrong_model","wrong_model_x")]

yax = c(readRDS("ref_tables/yax_aut.Rda"))
yax=yax[!names(yax) %in% coverage_sens_genes]
saveRDS(yax, "ref_tables/yax_aut_corr.Rda")

yax = c(readRDS("ref_tables/yax_xsa.Rda"))
yax=yax[!names(yax) %in% coverage_sens_genes]
saveRDS(yax, "ref_tables/yax_xsa_corr.Rda")

yax = c(readRDS("ref_tables/yax_par.Rda"))
yax=yax[!names(yax) %in% coverage_sens_genes]
saveRDS(yax, "ref_tables/yax_par_corr.Rda")

aut_post=read.table("ref_tables/autosome_hs.summary_stats.7_15.tsv",header=T)
aut_post$group = aut_post$group2 = "Autosome"
aut_post = aut_post[!aut_post$Gene %in% coverage_sens_genes,]
median(10^aut_post$log10_map)
mean(10^aut_post$log10_map)

weghorn=read.table("ref_tables/Weghorn_et_al_Supplementary_Table_1.txt",header=T)
test = merge(weghorn, aut_post, by="Gene")
cor.test(test$s_het_drift, 10^test$log10_map, method=c("spearman"))
cor.test(test$s_het_drift, 10^test$log10_map, method=c("pearson"))

aut_post=fread("ref_tables/autosome_hs.summary_stats.7_15.tsv",header=T)
aut_post$group = aut_post$group2 = "Autosome"
par_post = fread("ref_tables/par.summary_stats.6_9.tsv")
par_post$group = par_post$group2 = "X (PAR)"
y_homolog_gene_list = fread("ref_tables/npx_posteriors.9_13.txt")[,1]
colnames(y_homolog_gene_list) = "Gene"
y_homolog_gene_list$group2 = "X (Y-homolog, non-PAR)"
npx_post = read.table("ref_tables/avg_x.stats.6_1_22.tsv", header=TRUE)[,c("Gene","avg_hs_log10_ci_low", "avg_hs_log10_ci_high","avg_hs_log10_map")]
colnames(npx_post) = c("Gene","log10_ci_low", "log10_ci_high","log10_map")
npx_post$group = "non-par X"
x_post = merge(npx_post, y_homolog_gene_list, by="Gene", all.x=T)
x_post$group2[is.na(x_post$group2)] = "X (Other non-PAR)"
x_post = rbind(x_post, par_post)
all_post = rbind(x_post, aut_post)
all_post = all_post[!all_post$Gene %in% coverage_sens_genes,]
#all_post = all_post[!all_post$Gene %in% prob$gene[prob$flag=="duplicated"],]
gene_id_lookup = gnomad[,c("SYMBOL","Feature","mu_lof")][!is.na(mu_lof)][,c("SYMBOL","Feature")]
colnames(gene_id_lookup) = c("Gene","Ensembl_transcript_id")
write.table(gene_id_lookup, "out/gene_to_ensembl_id_map.txt",quote=F, sep="\t", row.names=F, col.names=T)
all_post = merge(all_post, gene_id_lookup, by="Gene", all.x=T)
write.table(all_post, "out/Supplementary_file_2.txt",quote=F, sep="\t", row.names=F, col.names=T)

```
### functions
```{r}
rlunif <- function(n, min, max, base=exp(1)) {
  if (mode(n) != "numeric")
    stop("'n' must be a non-empty numeric vector")
  if (any(missing(min), missing(max)))
    stop("'min' and 'max' not provided, without default.\n")
  ifelse(base == exp(1),
         return(exp(runif(n, log(min, base), log(max, base)))),
         return(base ^ (runif(n, log(min, base), log(max, base)))))
}

process_autosome_bootstrap <- function(p, label, group){

yax = c(readRDS("ref_tables/yax_aut_corr.Rda"))
xax = readRDS("ref_tables/xax.Rda")
iter = 1000
if (group==1){testcol="blue"} 
  else if (group==2){testcol="darkred"} 
  else if (group==3){testcol=wes_palette("GrandBudapest1")[4]}

mos <- setDT(p)[,c("gene","frac"),with=FALSE] 
colnames(mos) = c("gene","wt")
rem = data.frame(gene=names(yax)[!names(yax) %in% mos$gene])
rem$wt=0
mos = rbind(rem, mos)
mos=mos[match(names(yax), mos$gene),]
mos$wt = mos$wt/sum(mos$wt)

scale=as.list(mos$wt)
scale <- setNames(scale,mos$gene)
yax1=Map('*',yax,scale)
yax1=Reduce('+', yax1)
xax=as.numeric(unlist(xax))
  test_mean = weighted.mean(unlist(xax), yax1)
  f = approxfun(xax, yax1)
  C = try(integrate(f, min(unlist(xax)), max(unlist(xax)))$value)
  if(class(C) != "numeric") {
    C = integrate(f, min(unlist(xax)), max(unlist(xax)), subdivisions = 1000)$value
  }
  p.unscaled <- integrate(f, -1, max(unlist(xax)))$value
  p.scaled <- p.unscaled / C
  test_area=p.scaled
yax1 <- setNames(yax1,"y_axis")
post=list(xax,yax1)
post=data.frame(post[[1]], post[[2]])
colnames(post) = c("hs","density")
post$label="posterior"
post$group ="test"

post_all = readRDS("ref_tables/tmp_post_a.Rda")[label=="posterior"]
post = rbind(post, post_all)
post$col = "black"

post.ind=vector('list')
mean.ind = vector('list')
area.ind = vector('list')
for (i in 1:iter){
  mos = setDT(data.frame(sample(mos_all_a, sum(p$count), replace=TRUE)))
  colnames(mos) = c("gene")
  mos$count = 1  
  mos = mos[,list(count = sum(count)), by=list(gene)]
  mos$wt = mos$count/sum(mos$count) 
  mos$count=NULL
  mos$wt[is.na(mos$wt)]=0
  rem = data.frame(gene=names(yax)[!names(yax) %in% mos$gene])
  rem$wt=0
  mos = rbind(rem, mos)
  mos=mos[match(names(yax), mos$gene),]
  mos$wt = mos$wt/sum(mos$wt)
  
  scale=as.list(mos$wt)
  scale <- setNames(scale,mos$gene)
  yax1=Map('*',yax,scale)
  yax1=Reduce('+', yax1)
  xax=as.numeric(unlist(xax))
  mean.ind[[i]] = weighted.mean(xax, yax1)
  f <- approxfun(xax, yax1)
  C = try(integrate(f, min(unlist(xax)), max(unlist(xax)))$value)
  if(class(C) != "numeric") {
    C = integrate(f, min(unlist(xax)), max(unlist(xax)), subdivisions = 1000)$value
  }
  p.unscaled <- integrate(f, -1, max(unlist(xax)))$value
  p.scaled <- p.unscaled / C
  area.ind[[i]]=p.scaled
  yax1 <- setNames(yax1,"y_axis")
  tmp=list(xax,yax1)
  tmp=data.frame(tmp[[1]], tmp[[2]])
  colnames(tmp) = c("hs","density")
  tmp$group = "sampling distribution"
  post.ind[[i]]=tmp
}
post.ind = rbindlist(post.ind[1:100])
post.ind = rbind(post.ind, post, fill=T)
post.ind$group=factor(post.ind$group, levels = c("test","all possible DNMs (expectation)","sampling distribution"))

area.ind=as.numeric(unlist(area.ind))
rank=sum(as.numeric(test_area>area.ind))
pval_area = round(1-(rank/(1+iter)),3) #one sided

mean.ind=as.numeric(unlist(mean.ind))
rank=sum(as.numeric(test_mean>mean.ind))
pval_mean = round(ifelse(rank > iter/2, 2*(1-((rank)/(iter+1))),2*(rank)/(iter+1)),3) #two sided

test_plot=ggplot(data=post.ind, aes(x=10^hs, y=density, col=group)) + geom_point(size=1.2) +
    theme_bw() + xlab("hs") +
    theme(
    legend.position = "none", #c(.02, .99),
    legend.justification = c("left", "top"),
    legend.box.just = c("left", "top"),
    legend.margin = margin(1, 2, 2, 2),
    legend.title = element_blank(),
    legend.text = element_text(size=rel(0.9))
    ) + scale_color_manual(values=c(testcol,"black","lightgrey")) +
    scale_x_log10(limits = c(5e-7,1), breaks = trans_breaks("log10", function(x) 10^x),labels = trans_format("log10", math_format(10^.x))) +
  ylim(0,1.06) +  guides(color = guide_legend(override.aes = list(size=4, shape=15)))+
  annotate(geom="text",label=sprintf('%s',label),parse=FALSE,x=10^-6.25,y=0.99,size=4, col=testcol, hjust=0,fontface = "bold") +
  annotate(geom="text",label=paste0("Probability of hs>10% = ",round(test_area,2)),x=10^-6,y=0.35,size=3, col="black", hjust=0) +
  annotate(geom="text",label=paste0("p value = ",pval_mean),x=10^-6,y=0.45,size=3, col="black",hjust =0)

return(list(test_plot, test_area, pval_mean))
}

compare_cases = function(x,y){

iter=500
yax = c(readRDS("ref_tables/yax_aut_corr.Rda"))
xax = readRDS("ref_tables/xax.Rda")
gene.list = rep(as.vector(x$gene), times=as.numeric(x$count))
mean.ind.a = vector('list')
area.ind.a = vector('list')
post.ind=vector('list')
for (i in 1:iter){
  mos = setDT(data.frame(sample(gene.list, sum(x$count), replace=TRUE)))
  colnames(mos) = c("gene")
  mos$count = 1
  mos = mos[,list(count = sum(count)), by=list(gene)]
  mos$wt = mos$count/sum(mos$count)
  mos$count=NULL
  mos$wt[is.na(mos$wt)]=0
  rem = data.frame(gene=names(yax)[!names(yax) %in% mos$gene])
  rem$wt=0
  mos = rbind(rem, mos)
  mos=mos[match(names(yax), mos$gene),]

  scale=as.list(mos$wt)
  scale <- setNames(scale,mos$gene)
  yax1=Map('*',yax,scale)
  yax1=Reduce('+', yax1)
  xax=as.numeric(unlist(xax))
  mean.ind.a[[i]] = weighted.mean(xax, yax1)
  f <- approxfun(xax, yax1)
  C <- integrate(f, min(unlist(xax)), max(unlist(xax)))$value
  p.unscaled <- integrate(f, -1, max(unlist(xax)))$value
  p.scaled <- p.unscaled / C
  area.ind.a[[i]]=p.scaled
  yax1 <- setNames(yax1,"y_axis")
  tmp=list(xax,yax1)
  tmp=data.frame(tmp[[1]], tmp[[2]])
  colnames(tmp) = c("hs","density")
  tmp$group = "sampling dist"
  tmp$label = i
  post.ind[[i]]=tmp
}
post.ind = rbindlist(post.ind)
post.ind$group = "cases 1: sampling distribution"
cases_sampling = post.ind

yax = c(readRDS("ref_tables/yax_aut_corr.Rda"))
xax = readRDS("ref_tables/xax.Rda")
gene.list = rep(as.vector(y$gene), times=as.numeric(y$count))
post.ind=vector('list')
mean.ind.b = vector('list')
area.ind.b = vector('list')
for (i in 1:iter){
  mos = setDT(data.frame(sample(gene.list, sum(y$count), replace=TRUE)))
  colnames(mos) = c("gene")
  mos$count = 1
  mos = mos[,list(count = sum(count)), by=list(gene)]
  mos$wt = mos$count/sum(mos$count)
  mos$count=NULL
  mos$wt[is.na(mos$wt)]=0
  rem = data.frame(gene=names(yax)[!names(yax) %in% mos$gene])
  rem$wt=0
  mos = rbind(rem, mos)
  mos=mos[match(names(yax), mos$gene),]

  scale=as.list(mos$wt)
  scale <- setNames(scale,mos$gene)
  yax1=Map('*',yax,scale)
  yax1=Reduce('+', yax1)
  xax=as.numeric(unlist(xax))
  mean.ind.b[[i]] = weighted.mean(xax, yax1)
  f <- approxfun(xax, yax1)
  C <- integrate(f, min(unlist(xax)), max(unlist(xax)))$value
  p.unscaled <- integrate(f, -1, max(unlist(xax)))$value
  p.scaled <- p.unscaled / C
  area.ind.b[[i]]=p.scaled
  yax1 <- setNames(yax1,"y_axis")
  tmp=list(xax,yax1)
  tmp=data.frame(tmp[[1]], tmp[[2]])
  colnames(tmp) = c("hs","density")
  tmp$group = "sampling dist"
  tmp$label = i
  post.ind[[i]]=tmp
}
post.ind = rbindlist(post.ind[1:100])
post.ind$group = "cases 2: sampling distribution"
controls_sampling = post.ind

post.ind = rbind(cases_sampling,controls_sampling, fill=T)
post.ind$group=factor(post.ind$group, levels = c("cases 1: sampling distribution","cases 2: sampling distribution"))
post.ind$label[is.na(post.ind$label)]="other"

mean.ind.a=as.numeric(unlist(mean.ind.a))
mean.ind.b=as.numeric(unlist(mean.ind.b))
pval_mean = as.numeric(ks.test(mean.ind.a, mean.ind.b)[2])

df <- data.frame(
  a = mean.ind.a,
  b=mean.ind.b
)
compared_plot_mean = ggplot(df, aes(mean.ind.a)) + stat_ecdf(geom = "step") + stat_ecdf(aes(mean.ind.b), geom = "step", col="red")

area.ind.a=as.numeric(unlist(area.ind.a))
area.ind.b=as.numeric(unlist(area.ind.b))
pval_area = as.numeric(ks.test(area.ind.a, area.ind.b)[2])

df <- data.frame(
  a = area.ind.a,
  b=area.ind.b
)
compared_plot_area = ggplot(df, aes(area.ind.a)) + stat_ecdf(geom = "step") + stat_ecdf(aes(area.ind.b), geom = "step", col="red")



return(list(pval_mean, pval_area, compared_plot_mean, compared_plot_area))
}

# gene src file
# library(biomaRt)
# ensembl = useEnsembl(biomart="ensembl", dataset="hsapiens_gene_ensembl", mirror="www")
# ensembl.genes <- getBM(attributes=c('ensembl_gene_id','hgnc_symbol','hgnc_id','chromosome_name'), mart = ensembl)
# write.table(ensembl.genes, "ref_tables/ensembl_genes.txt",quote=F, sep="\t", row.names=F, col.names=T)
ensembl.genes = fread("ref_tables/ensembl_genes.txt", header=T)

# initialize supp table 3
sample_list = vector('list',20)

# area under dfe
yax = readRDS("ref_tables/yax_aut_corr.Rda")
xax = readRDS("ref_tables/xax.Rda")

# lookup gene mos
mos = fread("ref_tables/gnomad.v2.1.1.lof_metrics.by_gene.txt", header=T)
mos = mos[,c("gene","possible_lof","chromosome")]
mos = mos[gene %in% names(c(readRDS("ref_tables/yax_aut_corr.Rda"),readRDS("ref_tables/yax_xsa_corr.Rda")))]
mos = mos[!is.na(mos$possible_lof)]
mos_a = mos[!chromosome=="X"]
mos_all_a = rep(as.vector(mos_a$gene), times=as.numeric(mos_a$possible_lof))
mos_x = mos[chromosome=="X"]
mos_all_x = rep(as.vector(mos_x$gene[!is.na(mos_x$possible_lof)]), times=as.numeric(mos_x$possible_lof[!is.na(mos_x$possible_lof)]))

mos$wt = mos$possible_lof/sum(mos$possible_lof, na.rm=T)
mos = mos[,c("gene","wt")]
mos=mos[match(names(yax), mos$gene),]

scale=as.list(mos$wt)
scale <- setNames(scale,mos$gene)
yax=Map('*',yax,scale)
yax=Reduce('+', yax)
xax=as.numeric(unlist(xax))
  test_mean = weighted.mean(unlist(xax), yax)
  f <- approxfun(xax, yax)
  C <- integrate(f, min(unlist(xax)), max(unlist(xax)))$value
  p.unscaled <- integrate(f, -1, max(unlist(xax)))$value
  p.scaled <- p.unscaled / C
  test_area=p.scaled
  exp_area = test_area
  
```


## MAIN FIGS

### FIG 1
```{r}

#Make plot of hs distribution for autosomes
legend.col <- function(col, lev){
  opar <- par
  n <- length(col)
  bx <- par("usr")
  box.cx <- c(bx[2] + (bx[2] - bx[1]) / 1000,
              bx[2] + (bx[2] - bx[1]) / 1000 + (bx[2] - bx[1]) / 50)
  box.cy <- c(bx[3], bx[3])
  box.sy <- (bx[4] - bx[3]) / n
  xx <- rep(box.cx, each = 2)
  par(xpd = TRUE)
  for(i in 1:n){
    yy <- c(box.cy[1] + (box.sy * (i - 1)),
            box.cy[1] + (box.sy * (i)),
            box.cy[1] + (box.sy * (i)),
            box.cy[1] + (box.sy * (i - 1)))
    polygon(xx, yy, col = col[i], border = col[i])
  }
  par(new = TRUE)
  plot(0, 0, type = "n",
       ylim = c(min(lev), max(lev)),
       yaxt = "n", ylab = "",
       xaxt = "n", xlab = "",
       frame.plot = FALSE)
  axis(side = 4, las = 2, tick = FALSE, line = .25)
  par <- opar
}
z.plot1<-function(){
exp_posteriors<-read.table("ref_tables/autosome_hs.summary_stats.7_15.tsv",header=T)
exp_posteriors = exp_posteriors[!exp_posteriors$Gene %in% coverage_sens_genes,]
auto_P<-ecdf(exp_posteriors$log10_map)
exp_posteriors$ecdf<-auto_P(exp_posteriors$log10_map)
#Make plot for autosomes
plot(exp_posteriors$log10_map,exp_posteriors$ecdf,xlim=c(-7,0),cex=0.25,las=1,xaxt="n",xlab="Sex-averaged selection on loss of one copy",ylab="Cumulative Probability",ylim=c(0,1), cex.lab=1.4)
atx <- axTicks(1)
labels <- sapply(atx,function(i)
            as.expression(bquote(10^ .(i)))
          )
axis(1,at=atx,labels=labels, cex.lab=1.45)
title("Autosomes", adj = 0.05, line = -1.5, cex=2.5, font.main = 2)
rbPal <- colorRampPalette(c('blue','darkslategray1'))
exp_posteriors$ci_width<-exp_posteriors$log10_ci_high - exp_posteriors$log10_ci_low
ci_levels<-seq(6,0,-0.1)
exp_posteriors$Col <- rbPal(length(ci_levels))[as.numeric(cut(exp_posteriors$ci_width,breaks=ci_levels))]
segments(exp_posteriors$log10_ci_low,exp_posteriors$ecdf,exp_posteriors$log10_ci_high,lwd=.125,col=exp_posteriors$Col)
points(exp_posteriors$log10_map,exp_posteriors$ecdf, col="black", cex=0.25)
abline(v=-2,lty=4,col="darkgrey")
text(par("usr")[2]*3.5,mean(par("usr")[3:4])+0.45, as.expression(log[10]~"95% Credible Interval Width"), srt = -90, xpd = TRUE, pos = 4, cex=1.4)
legend.col(col = rbPal(length(ci_levels)), lev = ci_levels)
}

#Distribution of hs for X chromosome
z.plot2<-function(){
avg_x_param_posts<-read.table("ref_tables/avg_x.stats.6_1_22.tsv", header=TRUE)
avg_x_param_posts = avg_x_param_posts[!avg_x_param_posts$Gene %in% coverage_sens_genes,]
x_P<-ecdf(avg_x_param_posts$avg_hs_log10_map)
avg_x_param_posts$ecdf<-x_P(avg_x_param_posts$avg_hs_log10_map)
min(avg_x_param_posts$ecdf)
avg_x_param_posts$ci_width<-avg_x_param_posts$avg_hs_log10_ci_high - avg_x_param_posts$avg_hs_log10_ci_low
rbPal <- colorRampPalette(c('blue','darkslategray1'))
ci_levels<-seq(6,0,-0.1)
avg_x_param_posts$Col <- rbPal(length(ci_levels))[as.numeric(cut(avg_x_param_posts$ci_width,breaks=ci_levels))]
#Make plot for X
par(xpd=FALSE)
plot(avg_x_param_posts$avg_hs_log10_map,avg_x_param_posts$ecdf,xlim=c(-7,0),cex=0.25,las=1, xaxt="n",xlab="Sex-averaged selection on loss of one copy",ylab="Cumulative Probability",ylim=c(0,1), alpha=0.8, cex.lab=1.4)
title("X Chromosome", adj = 0.05, line = -1.5, cex=2, font.main = 2.5)
atx <- axTicks(1)
labels <- sapply(atx,function(i)
            as.expression(bquote(10^ .(i)))
          )
axis(1,at=atx,labels=labels, cex.lab=1.45)
segments(avg_x_param_posts$avg_hs_log10_ci_low,avg_x_param_posts$ecdf,avg_x_param_posts$avg_hs_log10_ci_high,lwd=.35,col=avg_x_param_posts$Col)
points(avg_x_param_posts$avg_hs_log10_map,avg_x_param_posts$ecdf, col="black", cex=0.25)
abline(v=-2,lty=4,col="darkgrey")
text(par("usr")[2]*3.5,mean(par("usr")[3:4])+0.45, as.expression(log[10]~"95% Credible Interval Width"), srt = -90, xpd = TRUE, pos = 4,cex=1.4)
legend.col(col = rbPal(length(ci_levels)), lev = ci_levels)
}


png(here("out/Fig1_a.png"),  height=6, width=14,units = "in", res=200)
par(mfrow = c(1, 2))
par(mar = c(5, 4, 3, 5.5))
z.plot1()
par(mar = c(5, 5.5, 3, 5))
z.plot2()
dev.off()


png1 = readPNG("out/Fig1_a.png")
png2 <- readPNG('out/Fig1_b.png')

png(here("out/Fig1.png"),  height=8, width=9.5,units = "in", res=200)
gt=grid.arrange(rasterGrob(png2, width = unit(8,"in"), height=unit(3.4,"in")),arrangeGrob(rasterGrob(png1), ncol=1),ncol=1, heights=c(1,1))
as_ggplot(gt) +
  draw_plot_label(label = c("A","B","C"), size = 15,
                  x = c(0, 0, 0.53), y = c(1, 0.5,0.5))
dev.off()



```

### FIG 2

#### DFE for autosomes only
```{r}

# DFE for autosomes only
yax = readRDS("ref_tables/yax_aut_corr.Rda")
xax = readRDS("ref_tables/xax.Rda")

# lookup gene mos
mos = fread("ref_tables/gnomad.v2.1.1.lof_metrics.by_gene.txt", header=T)
mos = mos[,c("gene","possible_lof","chromosome")]
mos = mos[!is.na(mos$possible_lof)]
mos = mos[!chromosome=="X"]
mos=mos[match(names(yax), mos$gene),]
mos$wt = mos$possible_lof/sum(mos$possible_lof, na.rm=T)
mos = mos[,c("gene","wt")]

scale=as.list(mos$wt)
scale <- setNames(scale,mos$gene)
yax=Map('*',yax,scale)
yax=Reduce('+', yax)
xax=as.numeric(unlist(xax))
  test_mean = weighted.mean(unlist(xax), yax)
  f <- approxfun(xax, yax)
  C <- integrate(f, min(unlist(xax)), max(unlist(xax)))$value
  p.unscaled <- integrate(f, -1, max(unlist(xax)))$value
  p.scaled <- p.unscaled / C
  test_area=p.scaled
  exp_area = test_area
yax <- setNames(yax,"y_axis")
ycdf = cumsum(as.numeric(yax))/sum(as.numeric(yax))
xcdf = as.numeric(unlist(xax))
post=list(xax,yax)
post=data.frame(post[[1]], post[[2]])
colnames(post) = c("hs","density")
post$label="posterior"

post=setDT(rbind(post))
post$group="all possible DNMs (expectation)"
saveRDS(post,"ref_tables/tmp_post_a.Rda")

fig2a=ggplot(data=post[label=="posterior"], aes(x=10^hs, y=density, col=group)) + geom_point(size=1.2) + 
    theme_bw() + xlab("hs") +
    theme(
    legend.position = "none", #c(.02, .99),
    legend.justification = c("left", "top"),
    legend.box.just = c("left", "top"),
    legend.margin = margin(1, 2, 2, 2),
    legend.title = element_blank(),
    legend.key.height = unit(0.6, 'cm'),
    legend.text = element_text(size=rel(0.7))) +
      #scale_color_brewer(palette=c("black","PuBuGn")) +
      scale_color_manual(values=c("black", c(brewer.pal(9, "PuBu")[c(8,7,6,5,4)]))) +
    scale_x_log10(limits = c(5e-7,1), breaks = trans_breaks("log10", function(x) 10^x),labels = trans_format("log10", math_format(10^.x))) +
  ylim(0,1) + guides(color = guide_legend(override.aes = list(size=2, shape=15))) +
  annotate(geom="text",label=sprintf("bold('All LOF mutational opportunities')"),parse=TRUE,x=10^-6.25,y=0.95,size=4, col="black", hjust=0)  
saveRDS(fig2a, file="out/fig2a.plot")


# numbers
t = approxfun(xcdf,ycdf)
1-t(-1)
1-t(-2)
```

#### DFE unascertained DNMs (goldmann)
```{r}
gc()
set.seed(100)
yax = c(readRDS("ref_tables/yax_aut_corr.Rda"))
xax = readRDS("ref_tables/xax.Rda")

load("ref_tables/dnm_src/final.dnm.Rda")

dnm = df[data %in% c("goldmann")]
dnm = dnm[Pheno %in% c("unascertained","control")]
dnm = dnm[!mut=="indel"]
nrow(dnm[canonical_csq %in% c("splice_donor_variant","splice_acceptor_variant","stop_gained")])/nrow(dnm)
dnm = dnm[canonical_csq %in% c("splice_donor_variant","splice_acceptor_variant","stop_gained")]

dnm$count=1
dnm = dnm[, list(count = sum(count)), by = list(SYMBOL, data)]
dnm$gene = dnm$SYMBOL

## goldmann
dnm.gene=dnm[data=="goldmann"]
dnm.gene$frac=dnm.gene$count/sum(dnm.gene$count)
mos = dnm.gene[,c("SYMBOL","frac")]
colnames(mos) = c("gene","wt")
rem = data.frame(gene=names(yax)[!names(yax) %in% mos$gene])
rem$wt=0
mos = rbind(rem, mos)
mos=mos[match(names(yax), mos$gene),]
mos$wt = mos$wt/sum(mos$wt)

scale=as.list(mos$wt)
scale <- setNames(scale,mos$gene)
yax1=Map('*',yax,scale)
yax1=Reduce('+', yax1)
xax=as.numeric(unlist(xax))
  test_mean = weighted.mean(unlist(xax), yax1)
  f <- approxfun(xax, yax1)
  C <- integrate(f, min(unlist(xax)), max(unlist(xax)))$value
  p.unscaled <- integrate(f, -1, max(unlist(xax)))$value
  p.scaled <- p.unscaled / C
  test_area=p.scaled
yax1 <- setNames(yax1,"y_axis")
post=list(xax,yax1)
post=data.frame(post[[1]], post[[2]])
colnames(post) = c("hs","density")
post$group ="Goldmann et al (37 dnms)"
post$col = "blue"
goldmann = post

iter=1000
post.ind=vector('list')
mean.ind = vector('list')
area.ind = vector('list')
for (i in 1:iter){
  mos = setDT(data.frame(sample(mos_all_a, sum(dnm$count), replace=TRUE)))
  colnames(mos) = c("gene")
  mos$count = 1  
  mos = mos[,list(count = sum(count)), by=list(gene)]
  mos$wt = mos$count/sum(mos$count) 
  mos$count=NULL
  mos$wt[is.na(mos$wt)]=0
  rem = data.frame(gene=names(yax)[!names(yax) %in% mos$gene])
  rem$wt=0
  mos = rbind(rem, mos)
  mos=mos[match(names(yax), mos$gene),]
  mos$wt = mos$wt/sum(mos$wt)
  
  scale=as.list(mos$wt)
  scale <- setNames(scale,mos$gene)
  yax1=Map('*',yax,scale)
  yax1=Reduce('+', yax1)
  xax=as.numeric(unlist(xax))
  mean.ind[[i]] = weighted.mean(xax, yax1)
  f <- approxfun(xax, yax1)
  C <- integrate(f, min(unlist(xax)), max(unlist(xax)), stop.on.error = FALSE)$value
  p.unscaled <- integrate(f, -1, max(unlist(xax)))$value
  p.scaled <- p.unscaled / C
  area.ind[[i]]=p.scaled
  yax1 <- setNames(yax1,"y_axis")
  tmp=list(xax,yax1)
  tmp=data.frame(tmp[[1]], tmp[[2]])
  colnames(tmp) = c("hs","density")
  tmp$group = "sampling_dist"
  post.ind[[i]]=tmp
}
post.ind = rbindlist(post.ind[1:100])
post_all = readRDS("ref_tables/tmp_post_a.Rda")[label=="posterior"]

post.ind = rbind(post.ind, post, post_all, fill=T)

area.ind=as.numeric(unlist(area.ind))
rank=sum(as.numeric(test_area>area.ind))
goldmann_pval_area = round(1-(rank/(1+iter)),3) #one sided

mean.ind=as.numeric(unlist(mean.ind))
rank=sum(as.numeric(test_mean>mean.ind))
goldmann_pval_mean = round(ifelse(rank > iter/2, 2*(1-((rank)/(iter+1))),2*(rank)/(iter+1)),3) #two sided

fig2b=ggplot(data=post.ind, aes(x=10^hs, y=density, col=group)) + geom_point(size=1.2) +
    theme_bw() + xlab("hs") +
    theme(
    legend.position = "none", #legend.position = c(.02, .99),
    legend.justification = c("left", "top"),
    legend.box.just = c("left", "top"),
    legend.margin = margin(1, 2, 2, 2),
    legend.title = element_blank(),
    legend.text = element_text(size=rel(0.7))
    ) + scale_color_manual(values=c("black","royal blue4","lightgrey")) +
    scale_x_log10(limits = c(5e-7,1), breaks = trans_breaks("log10", function(x) 10^x),labels = trans_format("log10", math_format(10^.x))) + ylim(0,1) +  
  guides(color = guide_legend(override.aes = list(size=2, shape=15))) + 
  annotate(geom="text",label=sprintf("bold('De novo LOF mutations in Goldmann et al. (800 trios)')"),parse=TRUE,x=10^-6.25,y=0.95,size=4, col="royal blue 4",hjust = 0)

saveRDS(fig2b, file="out/fig2b.plot")

#```

#### DFE unascertained DNMs (ssc)
#```{r}
gc()
set.seed(100)
yax = c(readRDS("ref_tables/yax_aut_corr.Rda"))
xax = readRDS("ref_tables/xax.Rda")

load("ref_tables/dnm_src/final.dnm.Rda")
dnm = df[data %in% c("ssc")]
dnm = dnm[Pheno %in% c("unascertained","control")]
dnm = dnm[!mut=="indel"]
dnm = dnm[canonical_csq %in% c("splice_donor_variant","splice_acceptor_variant","stop_gained")]

dnm$count=1
dnm = dnm[, list(count = sum(count)), by = list(SYMBOL, data)]
dnm$gene = dnm$SYMBOL

## ssc unaffected
# lookup gene mos
dnm.gene=dnm[data=="ssc"]
dnm.gene$frac=dnm.gene$count/sum(dnm.gene$count)
mos = dnm.gene[,c("SYMBOL","frac")]
colnames(mos) = c("gene","wt")
rem = data.frame(gene=names(yax)[!names(yax) %in% mos$gene])
rem$wt=0
mos = rbind(rem, mos)
mos=mos[match(names(yax), mos$gene),]
mos$wt = mos$wt/sum(mos$wt)

scale=as.list(mos$wt)
scale <- setNames(scale,mos$gene)
yax1=Map('*',yax,scale)
yax1=Reduce('+', yax1)
xax=as.numeric(unlist(xax))
  test_mean = weighted.mean(unlist(xax), yax1)
  f <- approxfun(xax, yax1)
  C <- integrate(f, min(unlist(xax)), max(unlist(xax)))$value
  p.unscaled <- integrate(f, -1, max(unlist(xax)))$value
  p.scaled <- p.unscaled / C
  test_area=p.scaled
yax1 <- setNames(yax1,"y_axis")
post=list(xax,yax1)
post=data.frame(post[[1]], post[[2]])
colnames(post) = c("hs","density")
post$group ="SSC unaffected sibs (64 DNMs)"
post$col = "green"
ssc_unaffected_sibs = post


iter=1000
post.ind=vector('list')
mean.ind = vector('list')
area.ind = vector('list')
for (i in 1:iter){
  mos = setDT(data.frame(sample(mos_all_a, sum(dnm$count), replace=TRUE)))
  colnames(mos) = c("gene")
  mos$count = 1  
  mos = mos[,list(count = sum(count)), by=list(gene)]
  mos$wt = mos$count/sum(mos$count) 
  mos$count=NULL
  mos$wt[is.na(mos$wt)]=0
  rem = data.frame(gene=names(yax)[!names(yax) %in% mos$gene])
  rem$wt=0
  mos = rbind(rem, mos)
  mos=mos[match(names(yax), mos$gene),]
  mos$wt = mos$wt/sum(mos$wt)
  
  scale=as.list(mos$wt)
  scale <- setNames(scale,mos$gene)
  yax1=Map('*',yax,scale)
  yax1=Reduce('+', yax1)
  xax=as.numeric(unlist(xax))
  mean.ind[[i]] = weighted.mean(xax, yax1)
  f <- approxfun(xax, yax1)
  C <- integrate(f, min(unlist(xax)), max(unlist(xax)))$value
  p.unscaled <- integrate(f, -1, max(unlist(xax)))$value
  p.scaled <- p.unscaled / C
  area.ind[[i]]=p.scaled
  yax1 <- setNames(yax1,"y_axis")
  tmp=list(xax,yax1)
  tmp=data.frame(tmp[[1]], tmp[[2]])
  colnames(tmp) = c("hs","density")
  tmp$group = "sampling_dist"
  post.ind[[i]]=tmp
}
post.ind = rbindlist(post.ind[1:100])
post.ind = rbind(post.ind, post, post_all, fill=T)

area.ind=as.numeric(unlist(area.ind))
rank=sum(as.numeric(test_area>area.ind))
ssc_controls_pval_area = round(1-(rank/(1+iter)),3) #one sided

mean.ind=as.numeric(unlist(mean.ind))
rank=sum(as.numeric(test_mean>mean.ind))
ssc_controls_pval_mean = round(ifelse(rank > iter/2, 2*(1-((rank)/(iter+1))),2*(rank)/(iter+1)),3) #two sided

fig2c=ggplot(data=post.ind, aes(x=10^hs, y=density, col=group)) + geom_point(size=1.2) +
    theme_bw() + xlab("hs") +
    theme(
    legend.position = "none", #c(.02, .99),
    legend.justification = c("left", "top"),
    legend.box.just = c("left", "top"),
    legend.margin = margin(1, 2, 2, 2),
    legend.key.height = unit(0.5, 'cm'),
    legend.title = element_blank(),
    legend.text = element_text(size=rel(0.7))
    ) + scale_color_manual(values=c("black","lightgrey","royal blue4","lightblue","navy","orange","pink")) +
    scale_x_log10(limits = c(5e-7,1), breaks = trans_breaks("log10", function(x) 10^x),labels = trans_format("log10", math_format(10^.x))) + ylim(0,1)+ 
  guides(color = guide_legend(override.aes = list(size=2, shape=15))) +
  annotate(geom="text",label=sprintf("bold('De novo LOF mutations in Simons Simplex siblings (1902 trios)')"),parse=TRUE,x=10^-6.25,y=0.95,size=4, col="royal blue 4", hjust=0)

fig2c
saveRDS(fig2c, file="out/fig2c.plot")

#```

#### sperm mutations (src supp table 5 in https://www.nature.com/articles/s41586-021-03822-7)
#```{r}
gc()
set.seed(100)
yax = c(readRDS("ref_tables/yax_aut_corr.Rda"))
xax = readRDS("ref_tables/xax.Rda")

test = fread("ref_tables/dnm_src/testes_somatic_mutations.txt.gz", header=T)
sample = fread("ref_tables/dnm_src/testes_somatic_sample.csv", header=T)
test = merge(test, sample, by="Sample")
test = test[test$ExonicFunc.refGene=="stopgain"]
test = test[,c("Gene.refGene")]
test$count=1
test$frac=test$count/sum(test$count)
test$gene = test$Gene.refGene
sample_n6 = test[,c("gene","count")]
sample_n6$data = "sperm"

mos = test[,c(1,3)]
colnames(mos) = c("gene","wt")
rem = data.frame(gene=names(yax)[!names(yax) %in% mos$gene])
rem$wt=0
mos = rbind(rem, mos)
mos=mos[match(names(yax), mos$gene),]
mos$wt = mos$wt/sum(mos$wt)

scale=as.list(mos$wt)
scale <- setNames(scale,mos$gene)
yax1=Map('*',yax,scale)
yax1=Reduce('+', yax1)
xax=as.numeric(unlist(xax))
  test_mean = weighted.mean(unlist(xax), yax1)
  f <- approxfun(xax, yax1)
  C <- integrate(f, min(unlist(xax)), max(unlist(xax)))$value
  p.unscaled <- integrate(f, -1, max(unlist(xax)))$value
  p.scaled <- p.unscaled / C
  test_area=p.scaled
yax1 <- setNames(yax1,"y_axis")
post=list(xax,yax1)
post=data.frame(post[[1]], post[[2]])
colnames(post) = c("hs","density")
post$group = "from mut sig paper (n=14)"
post$col = "red"
post_all = readRDS("ref_tables/tmp_post_a.Rda")[label=="posterior"]
post_all$label=NULL
post_all$col="black"
post=rbind(post_all, post)


iter=1000
post.ind=vector('list')
mean.ind = vector('list')
area.ind = vector('list')
for (i in 1:iter){
  mos = setDT(data.frame(sample(mos_all_a, sum(test$count), replace=TRUE)))
  colnames(mos) = c("gene")
  mos$count = 1  
  mos = mos[,list(count = sum(count)), by=list(gene)]
  mos$wt = mos$count/sum(mos$count) 
  mos$count=NULL
  mos$wt[is.na(mos$wt)]=0
  rem = data.frame(gene=names(yax)[!names(yax) %in% mos$gene])
  rem$wt=0
  mos = rbind(rem, mos)
  mos=mos[match(names(yax), mos$gene),]
  mos$wt = mos$wt/sum(mos$wt)
  
  scale=as.list(mos$wt)
  scale <- setNames(scale,mos$gene)
  yax1=Map('*',yax,scale)
  yax1=Reduce('+', yax1)
  xax=as.numeric(unlist(xax))
  mean.ind[[i]] = weighted.mean(xax, yax1)
  f <- approxfun(xax, yax1)
  C <- integrate(f, min(unlist(xax)), max(unlist(xax)), stop.on.error = FALSE)$value
  p.unscaled <- integrate(f, -1, max(unlist(xax)))$value
  p.scaled <- p.unscaled / C
  area.ind[[i]]=p.scaled
  yax1 <- setNames(yax1,"y_axis")
  tmp=list(xax,yax1)
  tmp=data.frame(tmp[[1]], tmp[[2]])
  colnames(tmp) = c("hs","density")
  tmp$group = "sampling_dist"
  post.ind[[i]]=tmp
}
post.ind = rbindlist(post.ind[1:100])
post.ind = rbind(post.ind, post, post_all, fill=T)

area.ind=as.numeric(unlist(area.ind))
rank=sum(as.numeric(test_area>area.ind))
sperm_pval_area = round(1-(rank/(1+iter)),3) #one sided

mean.ind=as.numeric(unlist(mean.ind))
rank=sum(as.numeric(test_mean>mean.ind))
sperm_pval_mean = round(ifelse(rank > iter/2, 2*(1-((rank)/(iter+1))),2*(rank)/(iter+1)),3) #two sided

fig2d=ggplot(data=post.ind, aes(x=10^hs, y=density, col=group)) + geom_point(size=1.2) +
    theme_bw() + xlab("hs") +
    theme(
    legend.position = "none", #c(.02, .99),
    legend.justification = c("left", "top"),
    legend.box.just = c("left", "top"),
    legend.margin = margin(1, 2, 2, 2),
    legend.title = element_blank(),
    legend.text = element_text(size=rel(0.7))
    ) + scale_color_manual(values=c("black","royal blue4","lightgrey")) +
    scale_x_log10(limits = c(5e-7,1), breaks = trans_breaks("log10", function(x) 10^x),labels = trans_format("log10", math_format(10^.x))) +
  ylim(0,1) +  guides(color = guide_legend(override.aes = list(size=2, shape=15)))+
    annotate(geom="text",label=sprintf("bold('Mutations observed in Sperm cells')"),parse=TRUE,x=10^-6.25,y=0.95,size=4, col="royal blue 4", hjust=0)

saveRDS(fig2d, file="out/fig2d.plot")

#```

#### plot
#```{r}

fig2a=readRDS("out/fig2a.plot")
fig2b=readRDS("out/fig2b.plot")
fig2c=readRDS("out/fig2c.plot")
fig2d=readRDS("out/fig2d.plot")

gt=grid.arrange(arrangeGrob(fig2a,fig2b, nrow=1), arrangeGrob(fig2c, fig2d, nrow=1), nrow=2)
p=as_ggplot(gt) +
  draw_plot_label(label = c("A", "B", "C","D"), size = 15,
                  x = c(0, .5, 0, .5), y = c(1, 1, 0.5, 0.5))
png(here("out/Fig2.png"),  height=8, width=12,units = "in", res=200)
print(p)
dev.off()

```


### FIG 3

#### ukb controls
```{r}

post = readRDS(post,file="ref_tables/plot.df.fig3b")
my_colors = c(c(brewer.pal(9, "PuBu")[c(8,7,6,5,4)]),"black")
names(my_colors) <- levels(post$group)

post$group2=expression(group)
fig3b=ggplot(data=post, aes(x=10^hs, y=density, col=group)) + geom_point(size=1.2) + 
    theme_bw() + xlab("hs") +
    theme(
    legend.position = c(.035, .88),
    legend.justification = c("left", "top"),
    legend.box.just = c("left", "top"),
    legend.margin = margin(1, 2, 2, 2),
    legend.title = element_blank(),
    legend.key.height = unit(0.6, 'cm'),
    legend.text = element_text(size=rel(0.7))) +
      #scale_color_brewer(palette=c("black","PuBuGn")) +
      scale_color_manual(values=my_colors,breaks = c('singletons', 'AC<10', 'AF<1e-3', 'AF<1e-2', 'AF<1e-1')) +
    scale_x_log10(limits = c(5e-7,1), breaks = trans_breaks("log10", function(x) 10^x),labels = trans_format("log10", math_format(10^.x))) +
  ylim(0,1) + guides(color = guide_legend(override.aes = list(size=2, shape=15)))+  annotate(geom="text",label=sprintf("bold('Segregating LOF variants in UK Biobank')"),parse=TRUE,x=10^-6.25,y=0.95,size=4, col="royal blue 4", hjust=0)

# invar sites ukb
post2 = readRDS(post,file="ref_tables/plot.df.fig3a")

fig3a=ggplot(data=post2, aes(x=10^hs, y=density, col=group)) + geom_point(size=1.2) + 
    theme_bw() + xlab("hs") +
    theme(
    legend.position = "none", #c(.02, .99),
    legend.justification = c("left", "top"),
    legend.box.just = c("left", "top"),
    legend.margin = margin(1, 2, 2, 2),
    legend.title = element_blank(),
    legend.key.height = unit(0.6, 'cm'),
    legend.text = element_text(size=rel(0.7))) +
      #scale_color_brewer(palette=c("black","PuBuGn")) +
      scale_color_manual(values=c("black", "#238A8DFF")) +
    scale_x_log10(limits = c(5e-7,1), breaks = trans_breaks("log10", function(x) 10^x),labels = trans_format("log10", math_format(10^.x))) +
  ylim(0,1) + guides(color = guide_legend(override.aes = list(size=2, shape=15)))+  annotate(geom="text",label=sprintf("bold('Genes with no observed LOF variants in UK Biobank')"),parse=TRUE,x=10^-6.25,y=0.95,size=4, col="#238A8DFF", hjust=0)

```

#### allele ages
```{r}

## sel sims 1
tmp.df = fread("ref_tables/sim_out/hs01.txt.gz")
colnames(tmp.df) = c("gen","aa","Aa","AA","non1","non2","non3","mu","sel","dom","gen_arose","xx")
tmp.df$popsize = tmp.df$aa + tmp.df$Aa + tmp.df$AA  #ind
tmp.df$maf = (tmp.df$aa + (0.5*tmp.df$Aa))/tmp.df$popsize
tmp.df = tmp.df[,c("mu","sel","dom","maf","popsize","gen_arose")]
tmp.df$hs = tmp.df$sel*tmp.df$dom
tmp.df$sample = 300000
tmp.df$acount = rbinom(n = nrow(tmp.df), size=tmp.df$sample, prob = tmp.df$maf)
hs_1perc = tmp.df[maf>0]
avg_af = mean(hs_1perc$acount/hs_1perc$sample)

## sel sims 2
tmp.df = fread("ref_tables/sim_out/hs10.txt.gz")
colnames(tmp.df) = c("gen","aa","Aa","AA","non1","non2","non3","mu","sel","dom","gen_arose","xx")
tmp.df$popsize = tmp.df$aa + tmp.df$Aa + tmp.df$AA  #ind
tmp.df$maf = (tmp.df$aa + (0.5*tmp.df$Aa))/tmp.df$popsize
tmp.df = tmp.df[,c("mu","sel","dom","maf","popsize","gen_arose")]
tmp.df$hs = tmp.df$sel*tmp.df$dom
tmp.df$sample = 300000
tmp.df$acount = rbinom(n = nrow(tmp.df), size=tmp.df$sample, prob = tmp.df$maf)
hs_10perc = tmp.df[maf>0]
avg_af = mean(hs_10perc$acount/hs_10perc$sample)

## sel sims 3
tmp.df = fread("ref_tables/sim_out/hs50.txt.gz")
colnames(tmp.df) = c("gen","aa","Aa","AA","non1","non2","non3","mu","sel","dom","gen_arose","xx")
tmp.df$popsize = tmp.df$aa + tmp.df$Aa + tmp.df$AA  #ind
tmp.df$maf = (tmp.df$aa + (0.5*tmp.df$Aa))/tmp.df$popsize
tmp.df = tmp.df[,c("mu","sel","dom","maf","popsize","gen_arose")]
tmp.df$hs = tmp.df$sel*tmp.df$dom
tmp.df$sample = 300000
tmp.df$acount = rbinom(n = nrow(tmp.df), size=tmp.df$sample, prob = tmp.df$maf)
hs_50perc = tmp.df[maf>0]
avg_af = mean(hs_50perc$acount/hs_50perc$sample)

tmp.df = rbind(hs_1perc, hs_10perc, hs_50perc)
tmp.df=tmp.df[,.SD[sample(.N, min(10000,.N))],by = hs]
tmp.df=tmp.df[,median:=median(gen_arose),by=list(hs)]
tmp.labels = unique(tmp.df[,c("hs","median")])
tmp.labels$label = paste0(tmp.labels$median," gen")

fig3c=ggplot(tmp.df, aes(gen_arose)) +
    geom_histogram(aes(y = ..count.. / sapply(PANEL, FUN=function(x) sum(count[PANEL == x]))), fill="darkgrey") +
    facet_wrap(~hs, nrow=3, ncol=1)+
  theme_bw() +
  scale_x_log10() +
  ylab("density") +
  xlab("Allele age if segregating at present (in generations)")+
  theme(
    legend.position = c(.01, .99),
    legend.justification = c("left", "top"),
    legend.box.just = c("left", "top"),
    legend.background = element_rect(colour = "white"),
    legend.margin = margin(2, 2, 2, 2),
    axis.text.x = element_text(angle=60, hjust=1),
    legend.text = element_text(size=rel(0.8)),
    legend.title = element_blank(),
    strip.background =element_rect(fill="white")) + theme(strip.text.x = element_blank()) +
  geom_text(data=tmp.labels, aes(label=tmp.labels$label,x=tmp.labels$median,y=0.5),size=3, col="darkred",hjust =-.45, inherit.aes = FALSE) +
  geom_vline(data=tmp.df, aes(xintercept = median, group = median), colour = 'darkred', lty="dashed") +
  geom_text(data=tmp.labels, aes(label=paste0("hs = ",hs),x=500,y=0.7),size=4, fontface="italic", col="darkred", inherit.aes = FALSE)


```

#### plot
```{r}

#plot

gt=grid.arrange(arrangeGrob(fig3a, fig3b, nrow=2), arrangeGrob(fig3c, nrow=1), nrow=1,widths=3:2)

p=as_ggplot(gt) +
  draw_plot_label(label = c("A", "B", "C"), size = 15,
                  x = c(0, 0, 0.6), y = c(1, 0.5, 1))


png(here("out/Fig3.png"),  height=8, width=12,units = "in", res=200)
print(p)
dev.off()

```
 

### FIG 4
#### DFE DDD (src EMS114546-supplement-SuppTable1 in Kaplanis et al)
```{r}
gc()
set.seed(100)

ddd = fread("ref_tables/dnm_src/ddd_dnms_all.txt", skip=0, header=T)
count_relevant_inds = length(unique(ddd$id))

syn = ddd[consequence %in% c("synonymous_variant")]
syn = syn[!symbol=="."]
syn$count=1
count_syn_mut = sum(syn$count)

ddd = ddd[consequence %in% c("stop_gained","splice_donor","splice_acceptor")]
ddd = ddd[!symbol=="."]
ddd = ddd[!nchar(alt)>1] 
ddd$count=1
ddd.gene = ddd[, list(count = sum(count)), by = list(symbol)]
ddd.gene$frac=ddd.gene$count/sum(ddd.gene$count)
ddd.gene$gene = ddd.gene$symbol
count_relevant_mut = sum(ddd.gene$count)

label = paste0('Developmental Disorders (',scales::comma(count_relevant_mut), ' DNMs in ',scales::comma(count_relevant_inds),' trios)')
assign(paste("fig4_p1",sep="_"),process_autosome_bootstrap(ddd.gene, label, 2))
saveRDS(fig4_p1[[1]], file="out/fig4_p1.plot")


# info for supp table
sample_list[[1]]=label
sample_list[[1]]$count_ind = count_relevant_inds
sample_list[[1]]$count_syn_per_ind =  count_syn_mut/count_relevant_inds
sample_list[[1]]$count_lof_per_ind =  count_relevant_mut/count_relevant_inds
sample_list[[1]]$prob_highly_del =  fig4_p1[[2]]
sample_list[[1]]$prob_causal =  (fig4_p1[[2]]-exp_area)/fig4_p1[[2]]
saveRDS(sample_list, file="sample_list")

```
#### DFE CHD https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5675000/ 
```{r}
gc()
set.seed(100)
sample_list = readRDS(file="sample_list")

chd = fread("ref_tables/dnm_src/chd_dnms.csv", skip=0, header=T)
count_relevant_inds = length(unique(chd$`Blinded ID`))

syn = chd[Variant_Class=="syn"]
syn = syn[!Gene=="."]
syn$count=1
count_syn_mut = sum(syn$count)

chd = chd[Variant_Class %in% c("non","splice")]
chd = chd[!Gene=="."]
chd = chd[!nchar(ALT)>1]
chd$count=1
chd.gene = chd[, list(count = sum(count)), by = list(Gene)]
chd.gene$frac=chd.gene$count/sum(chd.gene$count)
chd.gene$gene = chd.gene$Gene
count_relevant_mut = sum(chd.gene$count)

label = paste0('Congenital Heart Disease (',scales::comma(count_relevant_mut), ' DNMs in ',scales::comma(count_relevant_inds),' trios)')
assign(paste("fig4_p2",sep="_"),process_autosome_bootstrap(chd.gene, label, 2))
saveRDS(fig4_p2[[1]], file="out/fig4_p2.plot")

# info for supp table
sample_list[[2]]=label
sample_list[[2]]$count_ind = count_relevant_inds
sample_list[[2]]$count_syn_per_ind =  count_syn_mut/count_relevant_inds
sample_list[[2]]$count_lof_per_ind =  count_relevant_mut/count_relevant_inds
sample_list[[2]]$prob_highly_del =  fig4_p2[[2]]
sample_list[[2]]$prob_causal =  (fig4_p2[[2]]-exp_area)/fig4_p2[[2]]
saveRDS(sample_list, file="sample_list")


```
#### SEVERE epiLEPSY DNMs https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4185114/ (De novo mutations in synaptic transmission genes including DNM1 cause epileptic encephalopathies)
```{r}
gc()
set.seed(100)
sample_list = readRDS(file="sample_list")

epi1 = fread("ref_tables/dnm_src/epilepsy_dnms1.csv", header=T)
epi2 = fread("ref_tables/dnm_src/epilepsy_dnms2.csv", skip=1, header=T)
count_relevant_inds = sum(length(unique(epi1$`TRIO ID`)), length(unique(epi2$`trio ID`)))

syn1 = epi1[`Variant effect`=="synonymous"]
syn1$count=1
syn2 = epi2[`Coding variant type`=="synonymous_SNV"]
syn2$count=1
count_syn_mut = sum(syn1$count) + sum(syn2$count)

epi1 = epi1[epi1$`Variant effect` %in% c("stop gained", "splice donor","splice acceptor","frameshift")]
epi1$count = 1
epi1.gene = epi1[, .(count = sum(count)), by = list(Gene)]

epi2 = epi2[epi2$`Coding variant type` %in% c("stopgain_SNV", "canonical splice site_SNV")]
epi2$count = 1
epi2.gene = epi2[, .(count = sum(count)), by = list(`Gene Symbol`)]
colnames(epi2.gene) = c("Gene","count")

epi.gene = rbind(epi1.gene, epi2.gene)
epi.gene = epi.gene[, .(count = sum(count)), by = list(Gene)]
epi.gene$frac=epi.gene$count/sum(epi.gene$count)
epi.gene$gene = epi.gene$Gene
count_relevant_mut = sum(epi.gene$count)

label = paste0('Severe epilepsy (',scales::comma(count_relevant_mut), ' DNMs in ',scales::comma(count_relevant_inds),' trios)')
assign(paste("fig4_p3",sep="_"),process_autosome_bootstrap(epi.gene, label, 2))
saveRDS(fig4_p3[[1]], file="out/fig4_p3.plot")

# info for supp table
sample_list[[3]]=label
sample_list[[3]]$count_ind = count_relevant_inds
sample_list[[3]]$count_syn_per_ind =  count_syn_mut/count_relevant_inds
sample_list[[3]]$count_lof_per_ind =  count_relevant_mut/count_relevant_inds
sample_list[[3]]$prob_highly_del =  fig4_p3[[2]]
sample_list[[3]]$prob_causal =  (fig4_p3[[2]]-exp_area)/fig4_p3[[2]]
saveRDS(sample_list, file="sample_list")


```
#### DFE autism (ASC - mixed cohort) Satterstrom et al 2020 https://pubmed.ncbi.nlm.nih.gov/31981491/
```{r}
gc()
set.seed(100)

sample_list = readRDS(file="sample_list")

asc = fread("ref_tables/dnm_src/asc_dnm_full_list.csv", skip=0, header=T) #check if there are some multiplex families
ped = fread("ref_tables/dnm_src/asc_ped_file.csv", skip=0, header=T)
#ped = ped[grepl("ASC", Batch) |grepl("NIMH", Batch)]
asc = merge(asc, ped, by=c("Phenotype_ID"))
asc_n_fam = length(unique(asc$Family))
asc = asc[!GENE_NAME=="."]
asc = asc[Affected_Status.x==2]
#asc=asc[Child_Sex=="Male"]
count_relevant_inds = length(unique(asc$Child_ID))

syn = asc[VEP_functional_class_canonical_simplified=="synonymous_variant"]
syn$count=1
count_syn_mut = sum(syn$count)

asc = asc[VEP_functional_class_canonical_simplified %in% c("stop_gained","splice_donor_variant","splice_acceptor_variant")]
asc$count=1
asc.gene = asc[, list(count = sum(count)), by = list(GENE_NAME,Child_ID, Child_Sex)]
colnames(asc.gene) = c("gene","SampleID","sex","count")
asc.gene = asc.gene[, list(count = sum(count)), by = list(gene)]
asc.gene$frac=asc.gene$count/sum(asc.gene$count)
count_relevant_mut = sum(asc.gene$count)

label = paste0('Autism (',scales::comma(count_relevant_mut), ' DNMs in ',scales::comma(count_relevant_inds),' trios)')
assign(paste("fig4_p4",sep="_"),process_autosome_bootstrap(asc.gene, label, 2))
saveRDS(fig4_p4[[1]], file="out/fig4_p4.plot")

# info for supp table
sample_list[[4]]=label
sample_list[[4]]$count_ind = count_relevant_inds
sample_list[[4]]$count_syn_per_ind =  count_syn_mut/count_relevant_inds
sample_list[[4]]$count_lof_per_ind =  count_relevant_mut/count_relevant_inds
sample_list[[4]]$prob_highly_del =  fig4_p4[[2]]
sample_list[[4]]$prob_causal =  (fig4_p4[[2]]-exp_area)/fig4_p4[[2]]
saveRDS(sample_list, file="sample_list")

```
#### DFE (SCZ) combined Howrigan 2020 and Fromer 2014, Rees2020 and Xu 2012, with add info Supp table 1 https://www.nature.com/articles/nature12929#Sec10 and Supp Data 6 https://www.nature.com/articles/s41593-019-0564-3#Sec33
```{r}
gc()
set.seed(100)
sample_list = readRDS(file="sample_list")

scz1=fread("ref_tables/dnm_src/scz1_41586_2014_BFnature12929_MOESM39_ESM.csv", header=T)

ped=fread("ref_tables/dnm_src/scz2_taiw_SCZ_ped.csv", header=T)
scz2=fread("ref_tables/dnm_src/scz2_taiw_SCZ_dnm.csv", header=T)
scz2=merge(scz2, ped, by=c("TRIO_ID"), all.x=T)

scz3=fread("ref_tables/dnm_src/schizophrenia_dnms_3.csv", header=T)
scz4=fread("ref_tables/dnm_src/scz4_41586_2014_BFnature12929_MOESM40_ESM_part2.csv", header=T)[Study=="Xu"]
scz4 = scz4[scz4$`Child phenotype`=="SZ"]

count_relevant_inds = sum(length(unique(scz1$`Proband ID`)), length(unique(scz2$TRIO_ID)),265, 617)

syn1 = scz1[`Gene annotations`=="silent"]
syn1$count=1
syn2 = scz2[`ANNOTATION`=="synonymous"]
syn2$count=1
count_syn_mut = sum(syn1$count) + sum(syn2$count)

scz1$gene = scz1$Genes
scz1= scz1[scz1$`Gene annotations` %in% c("nonsense","esplice")]
scz1$count = 1
scz1 = scz1[,.(count=sum(count, na.rm=T)), by = list(gene)]

scz2$gene = scz2$GENE
scz2= scz2[scz2$ANNOTATION %in% c("ptv")]
scz2$count = 1
scz2 = scz2[,.(count=sum(count, na.rm=T)), by = list(gene)]

scz3$gene = scz3$Gene_Name
scz3= scz3[scz3$`Annotation` %in% c("stop_gained","Splice_SNV")]
scz3$count = 1
scz3 = scz3[,.(count=sum(count, na.rm=T)), by = list(gene)]

scz4$gene = scz4$Genes
scz4= scz4[scz4$`Gene annotations` %in% c("nonsense","esplice")]
scz4$count = 1
scz4 = scz4[,.(count=sum(count, na.rm=T)), by = list(gene)]

scz.gene = rbind(scz1, scz2, scz3, scz4)
scz.gene = scz.gene[,.(count=sum(count, na.rm=T)), by = list(gene)]
scz.gene=scz.gene[count>0]
scz.gene$frac=scz.gene$count/sum(scz.gene$count)
count_relevant_mut = sum(scz.gene$count)

label = paste0('Schizophrenia (',scales::comma(count_relevant_mut), ' DNMs in ',scales::comma(count_relevant_inds),' trios)')
assign(paste("fig4_p5",sep="_"),process_autosome_bootstrap(scz.gene, label, 2))
saveRDS(fig4_p5[[1]], file="out/fig4_p5.plot")

# info for supp table
sample_list[[5]]=label
sample_list[[5]]$count_ind = count_relevant_inds
sample_list[[5]]$count_syn_per_ind =  count_syn_mut/count_relevant_inds
sample_list[[5]]$count_lof_per_ind =  count_relevant_mut/count_relevant_inds
sample_list[[5]]$prob_highly_del =  fig4_p5[[2]]
sample_list[[5]]$prob_causal =  (fig4_p5[[2]]-exp_area)/fig4_p5[[2]]
saveRDS(sample_list, file="sample_list")

```
#### DFE Tourette/OCD https://pubmed.ncbi.nlm.nih.gov/31771860/ and https://www.cell.com/neuron/fulltext/S0896-6273(17)30350-1?_returnURL=https%3A%2F%2Flinkinghub.elsevier.com%2Fretrieve%2Fpii%2FS0896627317303501%3Fshowall%3Dtrue#supplementaryMaterial ### no info on fam history
```{r}
gc()
set.seed(100)
sample_list = readRDS(file="sample_list")

ocd1 = fread("ref_tables/dnm_src/tourette_dnms.csv", skip=0, header=T)
ocd1=ocd1[!source=="SSC"]
ocd2 = fread("ref_tables/dnm_src/ocd_dnms.csv", header=T)
ocd2=ocd2[batch=="ocd"]
count_relevant_inds = sum(length(unique(ocd1$CHILD)), length(unique(ocd2$Proband)))

syn1 = ocd1[`codingOverall`=="silent"]
syn1$count=1
syn2 = ocd2[`ExonicFunc.refGene`=="synonymous SNV"]
syn2$count=1
count_syn_mut = sum(syn1$count) + sum(syn2$count)

ocd1 = ocd1[nonsynType=="LGD"]
ocd1$symbol = ocd1$Gene.refGene
ocd1 = ocd1[,c("symbol")]
ocd2 = ocd2[Var=="LGD"]
ocd2$symbol = ocd2$Gene.refGene
ocd2 = ocd2[,c("symbol")]
ocd_cases = rbind(ocd1,ocd2)
ocd_cases$count=1
ocd.gene = ocd_cases[, list(count = sum(count)), by = list(symbol)]
ocd.gene$frac=ocd.gene$count/sum(ocd.gene$count)
ocd.gene$gene = ocd.gene$symbol
count_relevant_mut = sum(ocd.gene$count)

label = paste0('Tourette Syndrome/OCD (',scales::comma(count_relevant_mut), ' DNMs in ',scales::comma(count_relevant_inds),' trios)')
assign(paste("fig4_p6",sep="_"),process_autosome_bootstrap(ocd.gene, label,2))
saveRDS(fig4_p6[[1]], file="out/fig4_p6.plot")

# info for supp table
sample_list[[6]]=label
sample_list[[6]]$count_ind = count_relevant_inds
sample_list[[6]]$count_syn_per_ind =  count_syn_mut/count_relevant_inds
sample_list[[6]]$count_lof_per_ind =  count_relevant_mut/count_relevant_inds
sample_list[[6]]$prob_highly_del =  fig4_p6[[2]]
sample_list[[6]]$prob_causal =  (fig4_p6[[2]]-exp_area)/fig4_p6[[2]]
saveRDS(sample_list, file="sample_list")

```

#### repeats autism
```{r}
gc()
set.seed(100)

autism=asc.gene
asc.list = unique(as.character(autism$gene[autism$count > 0]))

### ALL SSC
autism.gene = autism[autism$count > 1]
autism.gene = autism.gene[,.(count=sum(count, na.rm=T)), by = list(gene)]
autism.gene$frac=autism.gene$count/sum(autism.gene$count)

label = paste0('Autism (',scales::comma(sum(autism.gene$count)), ' DNMs in ',scales::comma(nrow(autism.gene)),' recurrently hit genes)')
assign(paste("fig4_p7",sep="_"),process_autosome_bootstrap(autism.gene, label,2))
saveRDS(fig4_p7[[1]], file="out/fig4_p7.plot")

```

#### across cohorts [autism and DDD/CHD]
```{r}
gc()
set.seed(100)

ddd = rbind(ddd.gene, chd.gene, fill=T)[count>0]
ddd = ddd[,.(count=sum(count, na.rm=T)), by = list(gene)]
mixed.gene = merge(autism, ddd, by="gene")
mixed.gene$count = mixed.gene$count.x+mixed.gene$count.y
mixed.gene$count.x=NULL; mixed.gene$count.y=NULL

overlap.gene = mixed.gene
overlap.gene$frac=overlap.gene$count/sum(overlap.gene$count)

label = paste0('Autism and DD/CHD (',scales::comma(sum(overlap.gene$count)), ' DNMs in ',scales::comma(nrow(overlap.gene)),' shared genes)')
assign(paste("fig4_p8",sep="_"),process_autosome_bootstrap(overlap.gene, label,2))
saveRDS(fig4_p8[[1]], file="out/fig4_p8.plot")

```

#### across cohorts [autism and schizophrenia]
```{r}
gc()
set.seed(100)
autism=asc.gene[count>0]
schizophrenia = scz.gene[count>0]
schizophrenia = schizophrenia[,.(count=sum(count, na.rm=T)), by = list(gene)]
mixed.gene = merge(autism, schizophrenia, by="gene")
mixed.gene$count = mixed.gene$count.x+mixed.gene$count.y
mixed.gene$count.x=NULL; mixed.gene$count.y=NULL

overlap.gene = mixed.gene
overlap.gene$frac=overlap.gene$count/sum(overlap.gene$count)

label = paste0('Autism and Schizophrenia (',scales::comma(sum(overlap.gene$count)), ' DNMs in ',scales::comma(nrow(overlap.gene)),' shared genes)')
assign(paste("fig4_p9",sep="_"),process_autosome_bootstrap(overlap.gene, label,2))
saveRDS(fig4_p9[[1]], file="out/fig4_p9.plot")


```

#### compare diseases
```{r}
compare_cases(ddd.gene, scz.gene)
compare_cases(ddd.gene, ocd.gene)

```
#### plot
```{r}

fig4_p1=readRDS(file="out/fig4_p1.plot")
fig4_p2=readRDS("out/fig4_p2.plot")
fig4_p3=readRDS("out/fig4_p3.plot")
fig4_p4=readRDS("out/fig4_p4.plot")
fig4_p5=readRDS("out/fig4_p5.plot")
fig4_p6=readRDS("out/fig4_p6.plot")
fig4_p7=readRDS("out/fig4_p7.plot")
fig4_p8=readRDS("out/fig4_p8.plot")
fig4_p9=readRDS("out/fig4_p9.plot")

#plot
gt=grid.arrange(arrangeGrob(fig4_p1,fig4_p2,fig4_p3, nrow=1), arrangeGrob(fig4_p4,fig4_p5,fig4_p6, nrow=1), arrangeGrob(fig4_p7,fig4_p9,fig4_p8, nrow=1), nrow=3)

p=as_ggplot(gt) +
  draw_plot_label(label = c("A", "B", "C","D","E","F","G","H","I"), size = 15,
                  x = c(0, 0.33, 0.67, 0, 0.33, 0.67, 0, 0.33,0.67), y = c(1, 1, 1, 0.67, 0.67, 0.67, 0.33, 0.33, 0.33))


png(here("out/Fig4.png"),  height=12, width=18,units = "in", res=200)
print(p)
dev.off()

```



### FIG 5
#### DFE DDD subset w sex
```{r}
gc()
set.seed(100)
sample_list = readRDS(file="sample_list")

#males
ddd = fread("ref_tables/dnm_src/ddd_dnms_subset_w_sex.tab", skip=0, header=T)
ddd = ddd[sex =="M"]
count_relevant_inds = length(unique(ddd$new_id))

syn = ddd[consequence %in% c("synonymous_variant")]
syn = syn[!symbol=="."]
syn$count=1
count_syn_mut = sum(syn$count)

ddd = ddd[consequence %in% c("stop_gained","splice_donor","splice_acceptor")]
ddd = ddd[!symbol=="."]
ddd = ddd[!nchar(alt)>1]
ddd$count=1
ddd.gene = ddd[, list(count = sum(count)), by = list(symbol)]
ddd.gene$frac=ddd.gene$count/sum(ddd.gene$count)
ddd.gene$gene = ddd.gene$symbol
count_relevant_mut = sum(ddd.gene$count)
ddd.mal = ddd.gene

label = paste0('DDD (',scales::comma(count_relevant_mut), ' DNMs in ',scales::comma(count_relevant_inds),' affected males)')
assign(paste("fig5_p1",sep="_"),process_autosome_bootstrap(ddd.gene, label, 2))
saveRDS(fig5_p1[[1]], file="out/fig5_p1.plot")

# info for supp table
sample_list[[7]]=label
sample_list[[7]]$count_ind = count_relevant_inds
sample_list[[7]]$count_syn_per_ind =  count_syn_mut/count_relevant_inds
sample_list[[7]]$count_lof_per_ind =  count_relevant_mut/count_relevant_inds
sample_list[[7]]$prob_highly_del =  fig5_p1[[2]]
sample_list[[7]]$prob_causal =  (fig5_p1[[2]]-exp_area)/fig5_p1[[2]]

#females
ddd = fread("ref_tables/dnm_src/ddd_dnms_subset_w_sex_fromJK.tab", skip=0, header=T)
ddd = ddd[sex =="F"]
count_relevant_inds = length(unique(ddd$new_id))

syn = ddd[consequence %in% c("synonymous_variant")]
syn = syn[!symbol=="."]
syn$count=1
count_syn_mut = sum(syn$count)

ddd = ddd[consequence %in% c("stop_gained","splice_donor","splice_acceptor")]
ddd = ddd[!symbol=="."]
ddd = ddd[!nchar(alt)>1]
ddd$count=1
ddd.gene = ddd[, list(count = sum(count)), by = list(symbol)]
ddd.gene$frac=ddd.gene$count/sum(ddd.gene$count)
ddd.gene$gene = ddd.gene$symbol
count_relevant_mut = sum(ddd.gene$count)
ddd.fem = ddd.gene

label = paste0('DDD (',scales::comma(count_relevant_mut), ' DNMs in ',scales::comma(count_relevant_inds),' affected females)')
assign(paste("fig5_p2",sep="_"),process_autosome_bootstrap(ddd.gene, label, 3))
saveRDS(fig5_p2[[1]], file="out/fig5_p2.plot")

# info for supp table
sample_list[[8]]=label
sample_list[[8]]$count_ind = count_relevant_inds
sample_list[[8]]$count_syn_per_ind =  count_syn_mut/count_relevant_inds
sample_list[[8]]$count_lof_per_ind =  count_relevant_mut/count_relevant_inds
sample_list[[8]]$prob_highly_del =  fig5_p2[[2]]
sample_list[[8]]$prob_causal =  (fig5_p2[[2]]-exp_area)/fig5_p2[[2]]
saveRDS(sample_list, file="sample_list")

### compare males and females
#compare_cases(ddd.fem, ddd.mal)

```

#### DFE SSC w sex
```{r}
gc()
sample_list = readRDS(file="sample_list")

load("ref_tables/dnm_src/final.dnm.Rda")

## males
ssc = df[data=="ssc"]
ped = unique(ssc[,c("Fam","SampleID","sex_of_proband","Pheno")])
ped=ped[,n_in_fam:=.N, by=list(Fam)]
ped=ped[,sex_matched:=.N, by=list(Fam, sex_of_proband)]
ssc=ssc[Fam %in% ped$Fam]
ssc = ssc[Pheno=="case"]
ssc = ssc[sex_of_proband=="male"]
count_relevant_inds = length(unique(ssc$Fam))

syn = ssc[canonical_csq %in% c("synonymous_variant")]
syn$count=1
count_syn_mut = sum(syn$count)

ssc = ssc[ canonical_csq %in% c("stop_gained","splice_donor_variant","splice_acceptor_variant")]
ssc = ssc[!SYMBOL=="."]
ssc = ssc[!mut=="indel"]
ssc$count=1
ssc.gene = ssc[, list(count = sum(count)), by = list(SYMBOL)]
ssc.gene$frac=ssc.gene$count/sum(ssc.gene$count)
ssc.gene$gene = ssc.gene$SYMBOL
count_relevant_mut = sum(ssc.gene$count)

label = paste0('Simons Simplex (',scales::comma(count_relevant_mut), ' DNMs in ',scales::comma(count_relevant_inds),' affected males)')
assign(paste("fig5_p3",sep="_"),process_autosome_bootstrap(ssc.gene, label, 2))
saveRDS(fig5_p3[[1]], file="out/fig5_p3.plot")

# info for supp table
sample_list[[9]]=label
sample_list[[9]]$count_ind = count_relevant_inds
sample_list[[9]]$count_syn_per_ind =  count_syn_mut/count_relevant_inds
sample_list[[9]]$count_lof_per_ind =  count_relevant_mut/count_relevant_inds
sample_list[[9]]$prob_highly_del =  fig5_p3[[2]]
sample_list[[9]]$prob_causal =  (fig5_p3[[2]]-exp_area)/fig5_p3[[2]]

## females
ssc = df[data=="ssc"]
ped = unique(ssc[,c("Fam","SampleID","sex_of_proband","Pheno")])
ped=ped[,n_in_fam:=.N, by=list(Fam)]
ped=ped[,sex_matched:=.N, by=list(Fam, sex_of_proband)]
ssc=ssc[Fam %in% ped$Fam]
ssc = ssc[Pheno=="case"]
ssc = ssc[sex_of_proband=="female"]
count_relevant_inds = length(unique(ssc$Fam))

syn = ssc[canonical_csq %in% c("synonymous_variant")]
syn$count=1
count_syn_mut = sum(syn$count)

ssc = ssc[ canonical_csq %in% c("stop_gained","splice_donor_variant","splice_acceptor_variant")]
ssc = ssc[!mut=="indel"]
ssc$count=1
ssc.gene = ssc[, list(count = sum(count)), by = list(SYMBOL)]
ssc.gene$frac=ssc.gene$count/sum(ssc.gene$count)
ssc.gene$gene = ssc.gene$SYMBOL
count_relevant_mut = sum(ssc.gene$count)

label = paste0('Simons Simplex (',scales::comma(count_relevant_mut), ' DNMs in ',scales::comma(count_relevant_inds),' affected females)')
assign(paste("fig5_p4",sep="_"),process_autosome_bootstrap(ssc.gene, label, 3))
assign(paste("fig5_p4",sep="_"),process_autosome_bootstrap(ssc.gene, label, 3))
saveRDS(fig5_p4[[1]], file="out/fig5_p4.plot")

# info for supp table
sample_list[[10]]=label
sample_list[[10]]$count_ind = count_relevant_inds
sample_list[[10]]$count_syn_per_ind =  count_syn_mut/count_relevant_inds
sample_list[[10]]$count_lof_per_ind =  count_relevant_mut/count_relevant_inds
sample_list[[10]]$prob_highly_del =  fig5_p4[[2]]
sample_list[[10]]$prob_causal =  (fig5_p4[[2]]-exp_area)/fig5_p4[[2]]
saveRDS(sample_list, file="sample_list")

```

#### DFE dnm mssng spx w sex
```{r}
gc()
set.seed(100)
sample_list = readRDS(file="sample_list")

load("ref_tables/dnm_src/final.dnm.Rda")
mssng_samples1 = fread("ref_tables/dnm_src/mssng_samples_more_info.csv", header=T, skip=1)
colnames(mssng_samples1)[1] = "SampleID"
mssng_samples1 = mssng_samples1[ ,rel:=toString(RELATION), by=list(FAMILYID)]
mssng_samples2 = fread("ref_tables/dnm_src/mssng_samples.csv", header=T, skip=1)
colnames(mssng_samples2)[1] = "SampleID"
mssng_samples=merge(mssng_samples1, mssng_samples2, by="SampleID")

#males
mssng = df[data=="mssng2017"]
mssng = merge(mssng, mssng_samples, by="SampleID")
mssng = mssng[SEX=="M"]
mssng = mssng[!grepl("AffectedSibling",rel)]
count_relevant_inds=length(unique(mssng$FAMILYID))

syn = mssng[canonical_csq %in% c("synonymous_variant")]
syn$count=1
count_syn_mut = sum(syn$count)

#mssng = mssng[!mut=="indel"]
mssng = mssng[!Chr %in% c("chrX","chrY")]
mssng = mssng[canonical_csq %in% c("splice_donor_variant","splice_acceptor_variant","stop_gained")]
mssng$count=1
mssng.gene = mssng[, list(count = sum(count)), by = list(SYMBOL)]
mssng.gene$frac=mssng.gene$count/sum(mssng.gene$count)
mssng.gene$gene = mssng.gene$SYMBOL
count_relevant_mut = sum(mssng.gene$count)

label = paste0('MSSNG Simplex (',scales::comma(count_relevant_mut), ' DNMs in ',scales::comma(count_relevant_inds),' affected males)')
assign(paste("fig5_p5",sep="_"),process_autosome_bootstrap(mssng.gene, label, 2))
saveRDS(fig5_p5[[1]], file="out/fig5_p5.plot")

# info for supp table
sample_list[[11]]=label
sample_list[[11]]$count_ind = count_relevant_inds
sample_list[[11]]$count_syn_per_ind =  count_syn_mut/count_relevant_inds
sample_list[[11]]$count_lof_per_ind =  count_relevant_mut/count_relevant_inds
sample_list[[11]]$prob_highly_del =  fig5_p5[[2]]
sample_list[[11]]$prob_causal =  (fig5_p5[[2]]-exp_area)/fig5_p5[[2]]

#females
mssng = df[data=="mssng2017"]
mssng = merge(mssng, mssng_samples, by="SampleID")
mssng = mssng[SEX=="F"]
mssng = mssng[!grepl("AffectedSibling",rel)]
count_relevant_inds=length(unique(mssng$FAMILYID))

syn = mssng[canonical_csq %in% c("synonymous_variant")]
syn$count=1
count_syn_mut = sum(syn$count)

mssng = mssng[!Chr %in% c("chrX","chrY")]
mssng = mssng[canonical_csq %in% c("splice_donor_variant","splice_acceptor_variant","stop_gained")]

mssng$count=1
mssng.gene = mssng[, list(count = sum(count)), by = list(SYMBOL)]
mssng.gene$frac=mssng.gene$count/sum(mssng.gene$count)
mssng.gene$gene = mssng.gene$SYMBOL
count_relevant_mut = sum(mssng.gene$count)

label = paste0('MSSNG Simplex (',scales::comma(count_relevant_mut), ' DNMs in ',scales::comma(count_relevant_inds),' affected females)')
assign(paste("fig5_p6",sep="_"),process_autosome_bootstrap(mssng.gene, label, 3))
saveRDS(fig5_p6[[1]], file="out/fig5_p6.plot")

# info for supp table
sample_list[[12]]=label
sample_list[[12]]$count_ind = count_relevant_inds
sample_list[[12]]$count_syn_per_ind =  count_syn_mut/count_relevant_inds
sample_list[[12]]$count_lof_per_ind =  count_relevant_mut/count_relevant_inds
sample_list[[12]]$prob_highly_del =  fig5_p6[[2]]
sample_list[[12]]$prob_causal =  (fig5_p6[[2]]-exp_area)/fig5_p6[[2]]
saveRDS(sample_list, file="sample_list")

```

#### spark w sex
```{r}
gc()
set.seed(100)
sample_list = readRDS(file="sample_list")

# males
spark = fread("ref_tables/dnm_src/spark_pilot_dnms.csv", header=T)
spark = spark[`GENDER (1=male, 2=female)`==1]
count_relevant_inds = length(unique(spark$SPID))

syn = spark[Effect_cat %in% c("syn") & !nchar(ALT)>1]
syn$count=1
count_syn_mut = sum(syn$count)

spark = spark[Effect_cat=="lof"]
spark = spark[!nchar(ALT)>1]
spark = spark[,c("ANN.GENE")]
spark$count=1
spark.gene = spark[, list(count = sum(count)), by = list(ANN.GENE)]
spark.gene$frac=spark.gene$count/sum(spark.gene$count)
spark.gene$gene = spark.gene$ANN.GENE
count_relevant_mut = sum(spark.gene$count)

label = paste0('SPARK (',scales::comma(count_relevant_mut), ' DNMs in ',scales::comma(count_relevant_inds),' affected males)')
assign(paste("fig5_p7",sep="_"),process_autosome_bootstrap(spark.gene, label, 2))
saveRDS(fig5_p7[[1]], file="out/fig5_p7.plot")

# info for supp table
sample_list[[13]]=label
sample_list[[13]]$count_ind = count_relevant_inds
sample_list[[13]]$count_syn_per_ind =  count_syn_mut/count_relevant_inds
sample_list[[13]]$count_lof_per_ind =  count_relevant_mut/count_relevant_inds
sample_list[[13]]$prob_highly_del =  fig5_p7[[2]]
sample_list[[13]]$prob_causal =  (fig5_p7[[2]]-exp_area)/fig5_p7[[2]]

#females
spark = fread("ref_tables/dnm_src/spark_pilot_dnms.csv", header=T)
spark = spark[`GENDER (1=male, 2=female)`==2]
count_relevant_inds = length(unique(spark$SPID))

syn = spark[Effect_cat %in% c("syn")]
syn$count=1
count_syn_mut = sum(syn$count)

spark = spark[Effect_cat=="lof"]
spark = spark[!nchar(ALT)>1]
spark = spark[,c("ANN.GENE")]
spark$count=1
spark.gene = spark[, list(count = sum(count)), by = list(ANN.GENE)]
spark.gene$frac=spark.gene$count/sum(spark.gene$count)
spark.gene$gene = spark.gene$ANN.GENE
count_relevant_mut = sum(spark.gene$count)

label = paste0('SPARK (',scales::comma(count_relevant_mut), ' DNMs in ',scales::comma(count_relevant_inds),' affected females)')
assign(paste("fig5_p8",sep="_"),process_autosome_bootstrap(spark.gene, label, 3))
assign(paste("fig5_p8",sep="_"),process_autosome_bootstrap(spark.gene, label, 3)) # repaet if necessary
saveRDS(fig5_p8[[1]], file="out/fig5_p8.plot")

# info for supp table
sample_list[[14]]=label
sample_list[[14]]$count_ind = count_relevant_inds
sample_list[[14]]$count_syn_per_ind =  count_syn_mut/count_relevant_inds
sample_list[[14]]$count_lof_per_ind =  count_relevant_mut/count_relevant_inds
sample_list[[14]]$prob_highly_del =  fig5_p8[[2]]
sample_list[[14]]$prob_causal =  (fig5_p8[[2]]-exp_area)/fig5_p8[[2]]
saveRDS(sample_list, file="sample_list")

```

#### mssng mpx w sex
```{r}
gc()
set.seed(100)
sample_list = readRDS(file="sample_list")

load("ref_tables/dnm_src/final.dnm.Rda")
mssng_samples1 = fread("ref_tables/dnm_src/mssng_samples_more_info.csv", header=T, skip=1)
colnames(mssng_samples1)[1] = "SampleID"
mssng_samples1 = mssng_samples1[ ,rel:=toString(RELATION), by=list(FAMILYID)]
mssng_samples2 = fread("ref_tables/dnm_src/mssng_samples.csv", header=T, skip=1)
colnames(mssng_samples2)[1] = "SampleID"
mssng_samples=merge(mssng_samples1, mssng_samples2, by="SampleID")

# males
mssng = df[data=="mssng2017"]
mssng = merge(mssng, mssng_samples, by="SampleID")
mssng = mssng[SEX=="M"]
mssng = mssng[grepl("AffectedSibling",rel)]
count_relevant_inds=length(unique(mssng$FAMILYID))

syn = mssng[canonical_csq %in% c("synonymous_variant")]
syn$count=1
count_syn_mut = sum(syn$count)

mssng = mssng[!mut=="indel"]
mssng = mssng[canonical_csq %in% c("splice_donor_variant","splice_acceptor_variant","stop_gained")]
mssng$count=1
mssng.gene = mssng[, list(count = sum(count)), by = list(SYMBOL)]
mssng.gene$frac=mssng.gene$count/sum(mssng.gene$count)
mssng.gene$gene = mssng.gene$SYMBOL
count_relevant_mut = sum(mssng.gene$count)
mssng.mult.mal = mssng.gene

label = paste0('MSSNG Multiplex (',scales::comma(count_relevant_mut), ' DNMs in ',scales::comma(count_relevant_inds),' affected males)')
assign(paste("fig5_p9",sep="_"),process_autosome_bootstrap(mssng.gene, label, 2))
saveRDS(fig5_p9[[1]], file="out/fig5_p9.plot")

# info for supp table
sample_list[[15]]=label
sample_list[[15]]$count_ind = count_relevant_inds
sample_list[[15]]$count_syn_per_ind =  count_syn_mut/count_relevant_inds
sample_list[[15]]$count_lof_per_ind =  count_relevant_mut/count_relevant_inds
sample_list[[15]]$prob_highly_del =  fig5_p9[[2]]
sample_list[[15]]$prob_causal =  (fig5_p9[[2]]-exp_area)/fig5_p9[[2]]

#females
mssng = df[data=="mssng2017"]
mssng = merge(mssng, mssng_samples, by="SampleID")
mssng = mssng[SEX=="F"]
mssng = mssng[grepl("AffectedSibling",rel)]
count_relevant_inds=length(unique(mssng$FAMILYID))

syn = mssng[canonical_csq %in% c("synonymous_variant")]
syn$count=1
count_syn_mut = sum(syn$count)

mssng = mssng[!mut=="indel"]
mssng = mssng[canonical_csq %in% c("splice_donor_variant","splice_acceptor_variant","stop_gained")]
mssng$count=1
mssng.gene = mssng[, list(count = sum(count)), by = list(SYMBOL)]
mssng.gene$frac=mssng.gene$count/sum(mssng.gene$count)
mssng.gene$gene = mssng.gene$SYMBOL
count_relevant_mut = sum(mssng.gene$count)
mssng.mult.fem = mssng.gene

label = paste0('MSSNG Multiplex (',scales::comma(count_relevant_mut), ' DNMs in ',scales::comma(count_relevant_inds),' affected females)')
assign(paste("fig5_p10",sep="_"),process_autosome_bootstrap(mssng.gene, label, 3))
saveRDS(fig5_p10[[1]], file="out/fig5_p10.plot")

# info for supp table
sample_list[[16]]=label
sample_list[[16]]$count_ind = count_relevant_inds
sample_list[[16]]$count_syn_per_ind =  count_syn_mut/count_relevant_inds
sample_list[[16]]$count_lof_per_ind =  count_relevant_mut/count_relevant_inds
sample_list[[16]]$prob_highly_del =  fig5_p10[[2]]
sample_list[[16]]$prob_causal =  (fig5_p10[[2]]-exp_area)/fig5_p10[[2]]
saveRDS(sample_list, file="sample_list")

### compare males and females
compare_cases(mssng.mult.fem, mssng.mult.mal)

```

#### plot
```{r}

fig5_p1=readRDS(file="out/fig5_p1.plot")
fig5_p2=readRDS(file="out/fig5_p2.plot")
fig5_p3=readRDS("out/fig5_p3.plot")
fig5_p4=readRDS("out/fig5_p4.plot")
fig5_p5=readRDS("out/fig5_p5.plot")
fig5_p6=readRDS("out/fig5_p6.plot")
fig5_p7=readRDS("out/fig5_p7.plot")
fig5_p8=readRDS("out/fig5_p8.plot")
fig5_p9=readRDS("out/fig5_p9.plot")
fig5_p10=readRDS("out/fig5_p10.plot")

#plot
gt=grid.arrange(arrangeGrob(fig5_p1,fig5_p2, nrow=1), arrangeGrob(fig5_p3,fig5_p4, nrow=1), arrangeGrob(fig5_p5,fig5_p6, nrow=1), arrangeGrob(fig5_p7,fig5_p8, nrow=1), arrangeGrob(fig5_p9,fig5_p10, nrow=1), nrow=5)

p=as_ggplot(gt) +
  draw_plot_label(label = c("A", "B", "C","D","E","F","G","H","I","J"), size = 15,
                  x = c(0, 0.5, 0, 0.5, 0, 0.5, 0, 0.5, 0, 0.5), y = c(1, 1, 0.8, 0.8, 0.6, 0.6, 0.4, 0.4, 0.2, 0.2))


png(here("out/Fig5.png"),  height=18, width=12,units = "in", res=200)
print(p)
dev.off()

```






### Table S1
```{r echo=FALSE, results='asis', message=FALSE, warning=FALSE}            

sample_list = readRDS(file="sample_list")
test <- rbindlist(sample_list, fill=T)
colnames(test)[1] = "Sample"

#require(openxlsx)
list_of_datasets <- list("Supplementary_Table" = test)
write.table(list_of_datasets, file = "out/Table_S1.txt", quote=F, sep="\t",row.names = F, col.names = T)
```


